<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Задача" data-toc-modified-id="Задача-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Задача</a></span></li><li><span><a href="#Данные" data-toc-modified-id="Данные-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Данные</a></span></li><li><span><a href="#Метрика-качества" data-toc-modified-id="Метрика-качества-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Метрика качества</a></span></li><li><span><a href="#План-проекта" data-toc-modified-id="План-проекта-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>План проекта</a></span></li></ul></li><li><span><a href="#Импорт-данных" data-toc-modified-id="Импорт-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Определение-констант" data-toc-modified-id="Определение-констант-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Определение констант</a></span></li><li><span><a href="#Импорт-и-обзор-данных" data-toc-modified-id="Импорт-и-обзор-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Импорт и обзор данных</a></span></li></ul></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Стоп-слова" data-toc-modified-id="Стоп-слова-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Стоп-слова</a></span></li><li><span><a href="#Очистка-текста" data-toc-modified-id="Очистка-текста-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Очистка текста</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Разбиение-выборки-на-обучающую,-валидационную-и-тестовую" data-toc-modified-id="Разбиение-выборки-на-обучающую,-валидационную-и-тестовую-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Разбиение выборки на обучающую, валидационную и тестовую</a></span></li><li><span><a href="#Векторизация" data-toc-modified-id="Векторизация-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Векторизация</a></span><ul class="toc-item"><li><span><a href="#Мешок-слов-(BOW)" data-toc-modified-id="Мешок-слов-(BOW)-3.5.1"><span class="toc-item-num">3.5.1&nbsp;&nbsp;</span>Мешок слов (BOW)</a></span></li><li><span><a href="#TF-IDF" data-toc-modified-id="TF-IDF-3.5.2"><span class="toc-item-num">3.5.2&nbsp;&nbsp;</span>TF-IDF</a></span></li></ul></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Тестирование" data-toc-modified-id="Тестирование-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Тестирование</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Поиск токсичных пользовательских комментариев

## Описание проекта

### Задача

На основании имеющихся данных с комментариями пользователей необходимо подготовить прототип модели машинного обучения для классификации этих комментариев на токсичные и нетоксичные, что позволит в дальнейшем выявленые моделью негативные комментарии отправлять на модерацию. 
<br>Таким образом, перед нами задача бинарной классификации.

### Данные

В нашем распоряжении файл ***`toxic_comments.csv`***, содержащий набор данных с разметкой о токсичности представленных в нем комментариев. Файл содержит всего два столбца:
- *`text`* - собственно текст комментария
- *`toxic`* - признак того, что комментарий является токсичным

### Метрика качества

Согласно ТЗ в качестве метрики необходимо использовать $F_1$-меру. 

$$
F_1 = 2\cdot\frac{precision\cdot recall}{precision + recall}
$$

где &nbsp; $precision = \displaystyle\frac{TP}{TP + FP}$ &mdash; это точность, &nbsp; $recall = \displaystyle\frac{TP}{TP + FN}$ &mdash; это полнота, &nbsp; а $TP, FP$ &mdash; количество правильных и неверных предсказаний позитивного класса, &nbsp; $FN$ &mdash; количество неверных предсказаний негативного класса.

<br>Целевое значение метрики для прототипа &mdash; не ниже 0.75. 

### План проекта

1. Подготовка данных.
2. Исследовательский анализ данных.
3. Обучение моделей.
4. Выбор лучшей модели и проверка на тестовой выборке.
5. Выводы.

## Импорт данных

### Импорт библиотек

Импортируем необходимые библиотеки и функции.

In [1]:
import pandas as pd
from tqdm import tqdm
import nltk
import spacy
import sys
from nltk.corpus import stopwords as nltk_stopwords
from nltk import pos_tag
from nltk.tokenize import word_tokenize 
from sklearn.feature_extraction.text import TfidfVectorizer 
from sklearn.feature_extraction.text import CountVectorizer
import re
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, precision_score, recall_score, accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import GridSearchCV
from IPython.display import clear_output

In [2]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))
!{sys.executable} -m spacy download en_core_web_sm

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\LDV\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


     ---------------------------------------- 12.8/12.8 MB 9.2 MB/s eta 0:00:00
[+] Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


### Определение констант

In [3]:
RANDOM_STATE = 1234

### Импорт и обзор данных

Загрузим датасет и посмотрим его содержание.

In [4]:
df = pd.read_csv('./datasets/toxic_comments.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


In [5]:
df.head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

In [6]:
print('\033[1m Распределение положительного класса \033[0m')
pd.concat([df['toxic'].value_counts().rename('Кол-во'), (df['toxic'].value_counts() / len(df) * 100).apply(lambda x: '{:.2f}'.format(x)).rename('%')], axis=1)

 Распределение положительного класса 


Кол-во      %
0  143106  89.84
1   16186  10.16

Итак, имеем размеченный датасет с 159 571 комментариями. Пропусков нет. Положительный класс несбалансирован: датасет содержит всего 10% токсичных комментариев (*отметим, что положительный класс в нашем случае образован токсичными комметариями, для которых признак `toxic` равен 1*).

## Подготовка данных

### Стоп-слова

Сначала расширим список стоп-слов словами, которые несложно найти в корпусе после удаления стандартных стоп-слов из библиотеки nltk.

In [7]:
stopwords.update(["can't", "i'm", "that's", "'s", "i'd", "there's", "i've", "i'll", "don't", "they're", "arn't", "would", "won't", "that", "it", "no", "arn"])
print(sorted(stopwords))

["'s", 'a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', "aren't", 'arn', "arn't", 'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', "can't", 'couldn', "couldn't", 'd', 'did', 'didn', "didn't", 'do', 'does', 'doesn', "doesn't", 'doing', 'don', "don't", 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'hadn', "hadn't", 'has', 'hasn', "hasn't", 'have', 'haven', "haven't", 'having', 'he', 'her', 'here', 'hers', 'herself', 'him', 'himself', 'his', 'how', 'i', "i'd", "i'll", "i'm", "i've", 'if', 'in', 'into', 'is', 'isn', "isn't", 'it', "it's", 'its', 'itself', 'just', 'll', 'm', 'ma', 'me', 'mightn', "mightn't", 'more', 'most', 'mustn', "mustn't", 'my', 'myself', 'needn', "needn't", 'no', 'nor', 'not', 'now', 'o', 'of', 'off', 'on', 'once', 'only', 'or', 'other', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 're', 's', 'same', 'shan', "shan't", 'she', "she's",

### Очистка текста

Очистим текст, выполнив следующие шаги:
- переведём все тексты в нижний регистр
- заменим вхождение 3-х и более одинаковых букв на одну соответствующую букву
- удалим все символы кроме английских букв и апострофа
- удалим стоп-слова
- удалим все однобуквенные слова<br>

Для этого определим соответствующую функцию **clear_text**. Результат запишем в новый столбец *`clear_text`*.

In [8]:
def clear_text(text):
    text = re.sub(r'(\w)\1{2,}', r'\1', text.lower())
    text = ' '.join([word for word in re.sub(r"[^a-z ']", ' ', text).split()])
    text = ' '.join([word for word in text.split() if word not in stopwords])
    text = ' '.join([word for word in re.sub(r"[^a-z ]", ' ', text).split()])
    text = ' '.join([word for word in re.sub(r"\b\w{1}\b", ' ', text).split()])
    return text

In [9]:
tqdm.pandas()
df['clear_text'] = df['text'].progress_apply(clear_text)
print('\033[1m Комментарии после очистки \033[0m')
df.head()

100%|███████████████████████████████| 159292/159292 [00:13<00:00, 11632.40it/s]

 Комментарии после очистки 


text  toxic  \
0  Explanation\nWhy the edits made under my usern...      0   
1  D'aww! He matches this background colour I'm s...      0   
2  Hey man, I'm really not trying to edit war. It...      0   
3  "\nMore\nI can't make any real suggestions on ...      0   
4  You, sir, are my hero. Any chance you remember...      0   

                                          clear_text  
0  explanation edits made username hardcore metal...  
1  aww matches background colour seemingly stuck ...  
2  hey man really trying edit war guy constantly ...  
3  make real suggestions improvement wondered sec...  
4                      sir hero chance remember page

### Лемматизация

Лемматизацию выполним с помощью библиотеки *`spacy`*. Результат сохраним в колонке *`lemm_text`*.

In [10]:
nlp = spacy.load('en_core_web_sm', disable=['parser', 'ner'])
def spacy_lemmatize(text):
    doc = nlp(text)
    return " ".join([token.lemma_ for token in doc])

In [11]:
df['lemm_text'] = df['clear_text'].progress_apply(spacy_lemmatize)
print('\033[1m Тексты после лемматизации \033[0m')
df.head()

100%|█████████████████████████████████| 159292/159292 [14:59<00:00, 177.02it/s]

 Тексты после лемматизации 


text  toxic  \
0  Explanation\nWhy the edits made under my usern...      0   
1  D'aww! He matches this background colour I'm s...      0   
2  Hey man, I'm really not trying to edit war. It...      0   
3  "\nMore\nI can't make any real suggestions on ...      0   
4  You, sir, are my hero. Any chance you remember...      0   

                                          clear_text  \
0  explanation edits made username hardcore metal...   
1  aww matches background colour seemingly stuck ...   
2  hey man really trying edit war guy constantly ...   
3  make real suggestions improvement wondered sec...   
4                      sir hero chance remember page   

                                           lemm_text  
0  explanation edit make username hardcore metall...  
1  aww match background colour seemingly stick th...  
2  hey man really try edit war guy constantly rem...  
3  make real suggestion improvement wonder sectio...  
4                      sir hero chance remember page

### Разбиение выборки на обучающую, валидационную и тестовую

Разобьем датасет на обучающую и тестовую выборки в соотношении 4:1. Поскольку подбор гиперпараметров моделей будем осуществлять с помощью функции GridSearch и, соответственно, с использованием кросс-валидации - валидационную выборку выделять не будем.

In [12]:
df_train, df_test = train_test_split(df, test_size=0.2, random_state=RANDOM_STATE)
print(f'Размер обучающей выборки: {df_train.shape[0]}')
print(f'Размер тестовой выборки: {df_test.shape[0]}')

Размер обучающей выборки: 127433
Размер тестовой выборки: 31859


Проверим соотношение классов в полученных выборках.

In [13]:
print('\033[1m Соотношение классов в выборках (%)\033[0m')
pd.concat([(df_train['toxic'].value_counts() / len(df_train) * 100).rename('train'), (df_test['toxic'].value_counts() / len(df_test) * 100).rename('test')], axis=1)

 Соотношение классов в выборках (%)


train       test
0  89.797776  90.002825
1  10.202224   9.997175

Всё ок, баланс сохранился на уровне всего датасета.
<br>Теперь подготовим соответствующие корпусы текстов.

In [14]:
corpus_train = df_train['lemm_text'].values
corpus_test = df_test['lemm_text'].values

### Векторизация

Векторизацию выполним двумя простейшими способами: bow и tf_idf.

#### Мешок слов (BOW)

In [15]:
count_vect = CountVectorizer(stop_words=stopwords) 
count_vect.fit(corpus_train) 
bow_train = count_vect.transform(corpus_train)
bow_test = count_vect.transform(corpus_test)
print("Размер обучающей матрицы:", bow_train.shape)

Размер обучающей матрицы: (127433, 134065)


#### TF-IDF

In [16]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords) 
count_tf_idf.fit(corpus_train) 
tf_idf_train = count_tf_idf.transform(corpus_train) 
tf_idf_test = count_tf_idf.transform(corpus_test) 
print("Размер обучающей матрицы:", tf_idf_train.shape)

Размер обучающей матрицы: (127433, 134065)


## Обучение

Рассмотрим несколько простых моделей: логистическую регрессию, метод опорных векторов с линейным ядром и случайный лес. Подбор гиперпараметров моделей произведем с помощью GridSearch. В качестве метрики используем F1. Обучение проводим на двух обучающих матрицах (bow, tf-idf).

In [17]:
classifiers = [LinearSVC(), LogisticRegression(), RandomForestClassifier()]

param_grid = {     
    'LinearSVC': {
        'random_state': [RANDOM_STATE], 
        'class_weight': ['balanced'],
        'dual': [False],
        'C': [0.06]},     
    'LogisticRegression': {
        'random_state': [RANDOM_STATE], 
        'solver': ['lbfgs'],
        'C': [1.1],     
        'class_weight': ['balanced']},
    'RandomForestClassifier': {
        'random_state': [RANDOM_STATE],      
        'max_features': [100],
        'criterion': ["gini"], 
        'class_weight': ['balanced'],
        'n_estimators': [101],
        'max_depth': [1]}
}

In [18]:
%%time
best_estimators, best_scors, best_params = {}, {}, {}
for clf in classifiers:
    best_est, best_sc, best_par = [], [], []
    for i, vect_type in enumerate(['bow', 'tf-idf']):
        train = bow_train if i == 0 else tf_idf_train
        param = param_grid[type(classifiers[0]).__name__]
        grid = GridSearchCV(clf, param_grid[type(clf).__name__], cv=3, scoring='f1', verbose=0) 
        grid.fit(train, df_train['toxic'])  
        best_est.append(grid.best_estimator_)
        best_sc.append(grid.best_score_)
        best_par.append(grid.best_params_)
    best_estimators[type(clf).__name__] = best_est
    best_scors[type(clf).__name__] = best_sc
    best_params[type(clf).__name__] = best_par
clear_output()

Wall time: 6min 19s


Сравним полученные результаты.

In [19]:
print('\033[1m Таблица результатов, показанных моделями \033[0m')
pd.DataFrame(best_scors, index=['bow', 'tf-idf']).T

 Таблица результатов, показанных моделями 


bow    tf-idf
LinearSVC               0.757703  0.749730
LogisticRegression      0.759323  0.752038
RandomForestClassifier  0.247650  0.247323

Лучший результат у логистической регрессии на мешке слов.

## Тестирование

Поскольку лучшие результаты показала модель логистической регрессии на мешке слов, тестировать будем именно её.

In [20]:
predicted_test = best_estimators['LogisticRegression'][0].predict(bow_test)
print("F1 на тестовой выборке:", f1_score(df_test['toxic'], predicted_test))

F1 на тестовой выборке: 0.7627215551743854


Тестирование модели на тестовой выборке показало удовлетворительный результат, соответствующий заданным начальным условиям (F1 не менее 0.75). Посмотрим также на значения полноты и точности.

In [21]:
print('Значение Recall (полнота): {}'.format(recall_score(df_test['toxic'], predicted_test)))
print('Значение Precision (точность): {}'.format(precision_score(df_test['toxic'], predicted_test)))

Значение Recall (полнота): 0.8376766091051805
Значение Precision (точность): 0.7000787194961953


## Выводы

Целью проекта являлась разработка прототипа модели для бинарной классификации пользовательских комментариев на токсичные и нетоксичные. \

Размеченный датасет состоит из примерно 160 тысяч комментариев. В распределении целевого признака (токсичность текста) присутствует дисбаланс: положительный класс составляет всего 10% от размера выборки. Отдельно предобработка данных с целью уменьшения дисбаланса не проводилась, использовались штатные возможности моделей для учета этого фактора. \

Были выполнены следующие этапы обработки исходных текстов.
- Очистка текста от зашумляющих слов и символов. 
- Лемматизация оставшихся слов.
- Векторизация с использованием мешка слов bow и tf-idf.<br>

Подготовленный таким образом датасет был разбит на обучающую и тестовую выборки в соотношении 4:1. Подбор гиперпараметров выполнялся поиском по сетке, количество подвыборок для кросс-валидации принималось равным 3.

Наилучший результат $F_1 = 0.759$ показала модель логистической регрессии на мешке слов. Проверка на тестовой выборке доказала состоятельность данной модели, дав значение метрики $F_1 = 0.763$.<br>

Очевидно, что результат можно улучшить, проведя, например, более тщательную очистку текстов, или используя более совершенные модели для векторизации текстов и более мощные модели для обучения.